plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java

```diff
@@ -226,18 +284,98 @@ public void testMigrateUnsupportedColumnType()
+        assertQueryFails("SELECT count(*) FROM " + icebergTableName, "Not an Iceberg table: .*");
 
+        assertUpdate("DROP TABLE " + hiveTableName);
+    }
+
+    @DataProvider
+    public static Object[][] timestampPrecisionsAndValuesOrc()
+    {
+        return new Object[][] {
+                {HiveTimestampPrecision.MILLISECONDS, "2021-01-01 06:12:12.123"},
+                {HiveTimestampPrecision.MICROSECONDS, "2021-01-01 06:12:12.123456"}
+        };
+    }
+
+    @Test(dataProvider = "timestampPrecisionsAndValuesOrc")
+    public void testMigrateColumnTypeTimestampOnFormatOrc(HiveTimestampPrecision precision, String timestamp)
+    {
+        String tableName = "test_migrate_timestamp_orc" + randomNameSuffix();
+        String hiveTableName = "hive.tpch." + tableName;
+        String icebergTableName = "iceberg.tpch." + tableName;
+        Session session = timestampPrecisionSession(precision);
+
+        assertUpdate(session, "CREATE TABLE " + hiveTableName + " (x timestamp(" + precision.getPrecision() + ")) WITH (format='" + IcebergFileFormat.ORC + "') ");
+        assertUpdate(session, "INSERT INTO " + hiveTableName + " (x) values timestamp '" + timestamp + "'", 1);
+        assertQuery(session, "SELECT * FROM " + hiveTableName, "VALUES timestamp '" + timestamp + "'");
+        assertQueryFails(session, "SELECT * FROM " + icebergTableName, "Not an Iceberg table: .*");
+
+        assertQuerySucceeds(session, "CALL iceberg.system.migrate('tpch', '" + tableName + "')");
+
+        assertQuery(session, "SELECT * FROM " + icebergTableName, "VALUES timestamp '" + timestamp + "'");
+
+        assertUpdate(session, "DROP TABLE " + icebergTableName);
+    }
+
+    @Test
+    public void testMigrateColumnTypeTimestampMillisOnFormatAvro()
+    {
+        String tableName = "test_migrate_timestamp_millis_avro" + randomNameSuffix();
+        String hiveTableName = "hive.tpch." + tableName;
+        String icebergTableName = "iceberg.tpch." + tableName;
+        assertUpdate("CREATE TABLE " + hiveTableName + "  WITH (format='" + IcebergFileFormat.AVRO + "') AS SELECT timestamp '2021-01-01 00:00:00.000' x", 1);
         assertQuery("SELECT * FROM " + hiveTableName, "VALUES timestamp '2021-01-01 00:00:00.000'");
         assertQueryFails("SELECT * FROM " + icebergTableName, "Not an Iceberg table: .*");
 
+        assertQueryFails("CALL iceberg.system.migrate('tpch', '" + tableName + "')", "Timestamp precision \\(3\\) not supported for Iceberg with Avro file format. Use \"timestamp\\(6\\)\" instead.");
+
         assertUpdate("DROP TABLE " + hiveTableName);
     }
 
+    @DataProvider
+    public static Object[][] timestampPrecisionsAndValuesParquet()
+    {
+        return new Object[][] {
+                {HiveTimestampPrecision.MILLISECONDS, "2021-01-01 06:12:12.123", "2021-01-01 12:12:12.123"},
+                {HiveTimestampPrecision.MICROSECONDS, "2021-01-01 06:12:12.123456", "2021-01-01 12:12:12.123456"}
```

Comment:
Could you leave a code comment why these values are different? We may want to fix the behavior later (it's pre-existing though) and the comment will be helpful for the future code readers.

Reply:
Please file an issue.

Reply:
https://github.com/trinodb/trino/issues/17785